### PR TITLE
p2p: Protect last outbound HB compact block peer

### PR DESCRIPTION
**Pull request by sdaftuar**
If all our high-bandwidth compact block serving peers (BIP 152) stall block
download, then we can be denied a block for (potentially) a long time. As
inbound connections are much more likely to be adversarial than outbound
connections, mitigate this risk by never removing our last outbound HB peer if
it would be replaced by an inbound.

### CODE REVIEW

File: src/net_processing.cpp [hunk 1]
```diff
@@ -825,12 +825,27 @@ void PeerManagerImpl::MaybeSetPeerAsAnnouncingHeaderAndIDs(NodeId nodeid)
...
                 lNodesAnnouncingHeaderAndIDs.push_back(nodeid);
                 return;
             }
+            CNodeState *state = State(*it);
+            if (state != nullptr && !state->m_is_inbound) ++num_outbound_hb_peers;
+        }
+        if (nodestate->m_is_inbound) {
+            // If we're adding an inbound HB peer, make sure we're not removing
+            // our last outbound HB peer in the process.
+            if (lNodesAnnouncingHeaderAndIDs.size() >= 3 && num_outbound_hb_peers == 1) {
```
Comment (jonatack) [inline src/net_processing.cpp:841]: Optional, if retouch perhaps hoist the `3` here and in line 854 to a constexpr.

**Review by jonatack — COMMENTED** [review]
Concept and code review ACK 6efbcec4ded6116a42d2783c96c60ef0f255a1b2 

Running a node with this patch.

A test in `p2p_compactblocks.py` would be nice.

If this is implemented, should BIP152 "Implementation Notes" be updated with a note?

**Comment by Crypt-iQ** [comment]
ACK 6efbcec

As jonatack mentioned, I think extending the existing functional test would be good.

**Comment by jnewbery** [comment]
Concept ACK. It seems reasonable to protect one outbound peer as an HB compact block source.

I agree with @jonatack and @Crypt-iQ that it'd be good to see a test for this new behaviour.

I've always found the `MaybeSetPeerAsAnnouncingHeaderAndIDs()` logic to be strangely convoluted and difficult to follow - is there a good reason to recursively call into `ForNode()` instead of just calling it serially to first switch off hb for one node and then enable it for the other? If we did that, then this logic would be a lot easier to read - we wouldn't need to `std::swap` elements in a list only to later pop off the front of that list.

**Comment by sdaftuar** [comment]
I agree a test would be nice -- I'm not planning to write one myself, but if someone contributes one I'd be happy to include it here.

> I've always found the MaybeSetPeerAsAnnouncingHeaderAndIDs() logic to be strangely convoluted and difficult to follow - is there a good reason to recursively call into ForNode() instead of just calling it serially to first switch off hb for one node and then enable it for the other? If we did that, then this logic would be a lot easier to read - we wouldn't need to std::swap elements in a list only to later pop off the front of that list.

I don't know for sure, but my guess is that it's easiest to reason about the logic if the substitution of going from one HB peer to another all happens within one lock acquisition?  At any rate, I think the real problem with readability is having `ForNode()` at all; its existence implies that our model for how net and net_processing interact is broken (see eg #11604).  I think the goal is to eventually refactor all the code like this so that we don't have `ForNode()` or `ForEachNode()` in our code anymore, so I don't think there's much value in refactoring the existing logic here just to have it rewritten in the future.

**Comment by sipa** [comment]
Here is an RPC-based test for compact blocks HB selection: https://github.com/sipa/bitcoin/commits/sdaftuar-2021-06-reserve-outbound-hb

**Comment by sdaftuar** [comment]
@sipa Thanks, I've cherry-picked that commit onto this branch.

File: test/functional/p2p_compactblocks_hb.py [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+from test_framework.util import assert_equal
+
+
+class CompactBlocksConnectionTest(BitcoinTestFramework):
+    """Test class for verifying selection of HB peer connections."""
+
+    def set_test_params(self):
+        self.setup_clean_chain = True
+        self.num_nodes = 6
+        self.connection_ids = {}
```
Comment (jnewbery) [inline test/functional/p2p_compactblocks_hb.py:17]: This member variable is unused.

Comment (sipa) [reply test/functional/p2p_compactblocks_hb.py:17]: Oh, indeed. Leftover from an earlier iteration.

Comment (sipa) [reply test/functional/p2p_compactblocks_hb.py:17]: Made a few style improvements to the commit in https://github.com/sipa/bitcoin/commits/sdaftuar-2021-06-reserve-outbound-hb

Comment (sdaftuar) [reply test/functional/p2p_compactblocks_hb.py:17]: Thanks, cherry-picked the latest commit on that branch.

File: test/functional/p2p_compactblocks_hb.py [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+    """Test class for verifying selection of HB peer connections."""
+
+    def set_test_params(self):
+        self.setup_clean_chain = True
+        self.num_nodes = 6
+
+    def peer_info(self, from_node, to_node):
+        """Query from_node for its getpeerinfo about to_node."""
+        for peerinfo in self.nodes[from_node].getpeerinfo():
+            if "(testnode%i)" % to_node in peerinfo['subver']:
```
Comment (jnewbery) [inline test/functional/p2p_compactblocks_hb.py:21]: If you touch this branch again, consider using f-strings for string formatting (https://github.com/bitcoin/bitcoin/blob/master/test/functional/README.md#style-guidelines)

**Review by jonatack — COMMENTED** [review]
ACK 30aee2dfe671b347438c1c327c6f79edfacff1ce

Verified the test fails on master at line 81 (and at no other point) and passes on this branch; `pprint`ing the value of `status` at that point shows the following differences between master and this change:

```
master

[True,  True, False, True ]
[True,  True, True,  False]
[False, True, True,  True ]   <--- assert fails at this point

branch

[True, True,  False, True ]
[True, True,  True,  False]
[True, False, True,  True ]
```

<details><summary>A few suggestions for the test, feel free to ignore</summary><p>

```diff
@@ -18,7 +18,7 @@ class CompactBlocksConnectionTest(BitcoinTestFramework):
     def peer_info(self, from_node, to_node):
         """Query from_node for its getpeerinfo about to_node."""
         for peerinfo in self.nodes[from_node].getpeerinfo():
-            if "(testnode%i)" % to_node in peerinfo['subver']:
+            if f"(testnode{to_node})" in peerinfo['subver']:
                 return peerinfo
         return None
 
@@ -73,14 +73,13 @@ class CompactBlocksConnectionTest(BitcoinTestFramework):
         # Now relay one block through peer 2 (outbound from node 1), so it should take HB status
         # from one of the inbounds.
         status = self.relay_block_through(2)
-        assert_equal(status[0], True)
-        assert_equal(sum(status), 3)
+        assert_equal(status, [True, False, True, True])
 
         # Now relay again through nodes 3,4,5. Since 2 is outbound, it should remain HB.
         for nodeid in range(3, 6):
             status = self.relay_block_through(nodeid)
-            assert status[0]
-            assert status[nodeid - 2]
+            assert_equal(status[0], True)
+            assert_equal(status[nodeid - 2], True)
             assert_equal(sum(status), 3)
 
         # Reconnect peer 2, and retry. Now the three inbounds should be HB again.
@@ -88,8 +87,8 @@ class CompactBlocksConnectionTest(BitcoinTestFramework):
         self.connect_nodes(1, 2)
         for nodeid in range(3, 6):
             status = self.relay_block_through(nodeid)
-            assert not status[0]
-            assert status[nodeid - 2]
+            assert_equal(status[0], False)
+            assert_equal(status[nodeid - 2], True)
         assert_equal(status, [False, True, True, True])
```
</p></details>

(Very nice test.)

(Out of curiousity, do people see inbounds selected as HB peers?  Perhaps my peer topology has much better outbounds than inbounds, or not enough inbounds, but it's rare for even one inbound to provide a block and be selected for HB mode, much less three. Outbounds get the nod.)

**Comment by sipa** [comment]
My node at bitcoin.sipa.be currently has all 3 HB connections on inbounds (it has 250 connections).

**Comment by jonatack** [comment]
@sipa thanks -- good to hear about different peer topologies. Am probably restarting too often (for testing pulls) to see anything close to that many inbounds.

**Comment by achow101** [comment]
ACK 30aee2dfe671b347438c1c327c6f79edfacff1ce

Reviewed code and checked that the test fails on master but passes with this PR.

**Comment by ariard** [comment]
Code ACK 30aee2dfe

Our last oubound HB compact block peer could still be evicted under our CHAIN_SYNC_TIMEOUT/EXTRA_PEER_CHECK_INTERVAL logics if it appears slow on block-delivery. Hitting this eviction on our side would require from an attacker to control our peer's block-relay, assuming this peer's implementation has the same mitigations in place against adversarial peers, it sounds hard to achieve.

**Comment by naumenkogs** [comment]
Concept ACK. I agree with protecting last outbound HB peer.

**Pull request merged**